### PR TITLE
Plugins: Forward feature toggles to plugins

### PR DESCRIPTION
**Pull request by oshirohugo**
**What is this feature?**

Forward feature toggles enabled by setting plugins environment variables

**Why do we need this feature?**

This feature is part of the work necessary to allow backend plugins to have access to feature flags.

**Who is this feature for?**

This feature will be used by `grafana-plugin-sdk-go` to implement feature flags management.

**Which issue(s) does this PR fix?**:

This PR doesn't fully address it's related issue. A `grafana-plugin-sdk-go` part is still necessary.

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/plugins/envvars/envvars_test.go
Comment (xnyo) [inline pkg/plugins/envvars/envvars_test.go:364]: Nit: I would also add a test for multiple feature flags:

```suggestion
	})
	
	t.Run("backend datasource with multiple feature toggles", func(t *testing.T) {
		p := &plugins.Plugin{}
		envVarsProvider := NewProvider(&config.Cfg{
			Features: featuremgmt.WithFeatures("feat-1", true, "feat-2", true),
		}, nil)
		envVars, err := envVarsProvider.Get(context.Background(), p)
		require.NoError(t, err)
		assert.ElementsMatch(t, []string{"GF_VERSION=", "GF_FEATURE_TOGGLES_ENABLE=feat-1,feat-2"}, envVars)
	})
```

**Review by xnyo — COMMENTED** [review]
Great stuff!! :clap: :clap: :clap: 

I have left a couple of comments. I believe this is fine for now, but for remote plugin's sake, it may be worth checking how feature flags are used in core plugins right now, to see if we can easily move them to be per-request instead

(remote-plugins do not play well with env vars as they are not "processes" but rather a remote GRPC server, and there's an ongoing effort to make them multi-tenant)

**Review by xnyo — APPROVED** [review]
LGTM! But before merging this I would wait for @wbrowne 's review on this as well, regarding remote plugins, and also investigate how the feature flags are currently used in core plugins

**Comment by wbrowne** [comment]
> LGTM! But before merging this I would wait for @wbrowne 's review on this as well, regarding remote plugins, and also investigate how the feature flags are currently used in core plugins

Yeah it would be good if you could do some investigation @oshirohugo into whether it's possible to get away with having feature toggle state as part of the request or not (IE whether some core plugins are relying on feature toggle state before a request is made). It may not be what we want to do, but it would be good for our own understanding.

File: pkg/plugins/envvars/envvars.go
Comment (andresmgot) [inline pkg/plugins/envvars/envvars.go:100]: nit:

```suggestion
			for feat, i := range enabledFeatures {
				features[i] = feat
			}
```

**Review by andresmgot — APPROVED** [review]
IMO, we could go ahead with this approach for the moment and revisit this new environment variables with all the rest when/if we end up removing it.

File: pkg/plugins/envvars/envvars.go
Comment (oshirohugo) [reply pkg/plugins/envvars/envvars.go:100]: This suggestion won't work. `enabledFeatures` type is `map[string]bool`, therefore in the suggestion `i` would be `bool`.

Comment (andresmgot) [reply pkg/plugins/envvars/envvars.go:100]: ah, right, right, ignore this. The other possibility is define slice with capacity and append instead:

```go
			features := make([]string, 0, len(enabledFeatures))
			for feat := range enabledFeatures {
					features = append(features, feat)
			}
```

but it's the same, just a matter of style, feel free to ignore.

File: pkg/plugins/ifaces.go
Comment (ryantxu) [inline pkg/plugins/ifaces.go:167]: is this required?  if yes, could/should it be an array?  (do we send false values?)

**Comment by ryantxu** [comment]
This seems like a reasonable first step, but I am a little worried about the impact on efforts to support multi-tenancy 😬 cc @wbrowne 

an an alternative solution may be to forward a set of flags (maybe prefix or explicitly configured) along in the gRPC context with each request, and then populated in the execution context?

This would shift feature toggles from static init to context dependant -- likely good
  
alternatively we may just need to ensure that multi-tenant plugins are only shared between tenants with the same feature flags 🤔

File: pkg/plugins/ifaces.go
Comment (ryantxu) [reply pkg/plugins/ifaces.go:167]: Since this is all internal, it maybe does not matter

**Comment by wbrowne** [comment]
> This seems like a reasonable first step, but I am a little worried about the impact on efforts to support multi-tenancy 😬 cc @wbrowne
> 
> an an alternative solution may be to forward a set of flags (maybe prefix or explicitly configured) along in the gRPC context with each request, and then populated in the execution context?
> 
> This would shift feature toggles from static init to context dependant -- likely good
> 
> alternatively we may just need to ensure that multi-tenant plugins are only shared between tenants with the same feature flags 🤔

👋 Thanks for looking out @ryantxu. I'm aware of these changes and we've chatted about them amongst the team. This is very much a short term thing to unblock the core plugins work. @stephaniehingtgen and I have thrown together a doc that takes a stab at how we solve for all configuration like this. The plan is to get going on a solution ASAP. Will it share with you.

File: pkg/plugins/ifaces.go
Comment (oshirohugo) [reply pkg/plugins/ifaces.go:167]: In this interface I'm only exposing a method that already exists in `FeatureManger`
https://github.com/grafana/grafana/blob/main/pkg/services/featuremgmt/manager.go#L132

If I want to return an array I would have to include a new method in `FeatureManager`

File: pkg/plugins/envvars/envvars.go
Comment (andresmgot) [reply pkg/plugins/envvars/envvars.go:92]: didn't you change this to use a proper context?

**Review by andresmgot — APPROVED** [review]
It seems that you are getting a a11y issue due to the GPG rotation from yesterday, I am not sure who owns that chech though, you can try in #grafana-delivery

**Pull request merged**
